mlflow/genai/optimize/adapt.py

```diff
@@ -0,0 +1,181 @@
+
+    # TODO: Add dataset validation
+    converted_train_data = _convert_eval_set_to_df(train_data).to_dict("records")
+    predict_fn = convert_predict_fn(
+        predict_fn=predict_fn, sample_input=converted_train_data[0]["inputs"]
+    )
+
+    eval_fn = _build_eval_fn(predict_fn)
+
+    target_prompts = [load_prompt(prompt_uri) for prompt_uri in target_prompt_uris]
+    target_prompts_dict = {prompt.name: prompt.template for prompt in target_prompts}
+
+    optimizer_output = optimizer.optimize(
+        eval_fn, converted_train_data, target_prompts_dict, optimizer_lm_params
+    )
+
+    return PromptAdaptationResult(
+        optimized_prompts=[
+            register_prompt(name=prompt_name, template=prompt)
+            for prompt_name, prompt in optimizer_output.optimized_prompts.items()
+        ],
+        optimizer_name=optimizer.__class__.__name__,
+    )
+
+
+def _build_eval_fn(
+    predict_fn: Callable[..., Any],
+) -> Callable[[dict[str, str], list[dict[str, Any]]], list[EvaluationResultRecord]]:
+    """
+    Build an evaluation function that uses the candidate prompts to evaluate the predict_fn.
+    """
+    from mlflow.pyfunc import Context, set_prediction_context
+
+    def eval_fn(
+        candidate_prompts: dict[str, str], dataset: list[dict[str, Any]]
+    ) -> list[EvaluationResultRecord]:
+        @property
+        def _template_patch(self) -> str:
+            template_name = self.name
+            if template_name in candidate_prompts:
+                return candidate_prompts[template_name]
+            return self.template
+
+        patch = _wrap_patch(PromptVersion, "template", _template_patch)
+
+        def _run_single(record: dict[str, Any]):
+            inputs = record["inputs"]
+            outputs = record["outputs"]
+            eval_request_id = str(uuid.uuid4())
+            # set prediction context to retrieve the trace by the request id,
+            # and set is_evaluate to True to disable async trace logging
+            with set_prediction_context(Context(request_id=eval_request_id, is_evaluate=True)):
+                try:
+                    target_outputs = predict_fn(inputs)
+                except Exception as e:
+                    target_outputs = f"Failed to invoke the predict_fn with {inputs}: {e}"
+
+            trace = mlflow.get_trace(eval_request_id, silent=True)
+            # TODO: Consider more robust scoring mechanism
+            score = 1 if target_outputs == outputs else 0
+            return EvaluationResultRecord(
+                inputs=inputs, outputs=target_outputs, score=score, trace=trace
+            )
+
+        try:
+            with ThreadPoolExecutor(
+                max_workers=MLFLOW_GENAI_EVAL_MAX_WORKERS.get(),
+                thread_name_prefix="MlflowGenAIEvalHarness",
+            ) as executor:
+                futures = [executor.submit(_run_single, record) for record in dataset]
+                return [future.result() for future in futures]
```

Comment:
nonblocking - We may want to use the lazy schedule instead of scheduling all at once to reduce the scheduling overhead:

```
with ThreadPoolExecutor(...) as executor:
    return list(executor.map(_run_single, dataset))
```